Support ASCI control

johnwparent opened the pull request:
Allow Spack to set Windows console associated with Spack process to handle ASCI control characters. This allows for colorization on Windows and removed the significant output cruft that appears in STDOUT when this functionality is not enabled. Behavior is enabled as soon as the tty module is loaded.

johnwparent committed:
Support ASCI control

Windows CMD prompt does not automatically support ASCI color control
characters on the console from Python. Enable this behavior by
accessing the current console and allowing the interpreation of ASCI
control characters from Python via the win32 API.

scheibelp commented on lib/spack/llnl/util/tty/color.py:
```diff
@@ -107,6 +107,56 @@ def __init__(self, message):
+# on by default. Turn on before we try to write to console
+# with color
+if sys.platform == "win32":
+    _enable_windows_vtp()
```
Would there be cases where this fails (and I assume one of the control commands in this function would raise an exception)? This could catch it, set `tty.color._force_color = False`, and then proceed (possibly also printing a debug message with the exception).

johnwparent replied:
You're right, ostensibly any of these commands could fail, and I have no idea how they'll behave on Windows 11, so I'll add the error handling.

johnwparent replied:
Addressed.

johnwparent commented:
@scheibelp I believe I've addressed all the changes requested during out meeting yesterday.

johnwparent commented:
@scheibelp RE: our last conversation about this pr, looks like I was wrong and we do have to enable this color functionality each time spack is executed, rather than just once in the startup script unfortunately. I've refactored to reflect that, and folded in some changes I made in the course of testing the discussed functionality.

johnwparent committed:
Update haspywin

Inform user Spack is installing pyWin32 module behind the scenes.
This is neccesary for Windows support.

johnwparent committed:
Remove Spack external find from startup

Spack was running an external detection of Python during each invocation
of the setup script for Windows CMD/PWSH, which has dramatic performance
implications each time the script is invoked, and is completely
unneccesary. Remove this operation.
